### PR TITLE
Docs/describe request middleware

### PR DESCRIPTION
**Pull request by EugeniyBurmistrov**

### CODE REVIEW

**Review by max-vasin — APPROVED** [review]

File: docs/api-reference.md [hunk 1]
```diff
@@ -1730,3 +1730,235 @@ Resolves a relative path to an absolute URL within the application.
...
+        ...
+      ]
+    }
+  },
+  (error, result) => {
+    ...
+  }
+```
+
+If you use multiple middleware functions, they run in the order that they are specified in the options object.
```
Comment (dirk-pieterse) [inline docs/api-reference.md:1803]:
```suggestion
Multiple middleware functions are run in the order they are specified in the options object.
```

File: docs/api-reference.md [hunk 2]
```diff
@@ -1730,3 +1730,235 @@ Resolves a relative path to an absolute URL within the application.
...
+// Export the factory function.
+export const createMyMiddleware = (options) =>
+  waitForResponse.bind(null, options)
+```
+
+The `params` object exposes the following API:
+
+| Field Name   | Description                                                                      |
+| ------------ | -------------------------------------------------------------------------------- |
+| fetch        | A JavaScript fetch function that you can use to perform arbitrary HTTP requests. |
```
Comment (dirk-pieterse) [inline docs/api-reference.md:1957]:
```suggestion
| fetch        | A JavaScript fetch function you can use to perform arbitrary HTTP requests. |
```

File: docs/api-reference.md [hunk 3]
```diff
@@ -1730,3 +1730,235 @@ Resolves a relative path to an absolute URL within the application.
...
+| info         | An object that describes the current request.                                    |
+| init         | An object that is the fetch function's `init` parameter.                         |
+| repeat       | A function that you can call to repeat the current request.                      |
+| end          | Call this function to commit the middleware execution result or error.           |
+| state        | A state object passed between middleware functions.                              |
+| deserializer | Given a string, returns a deserealized object.                                   |
```
Comment (dirk-pieterse) [inline docs/api-reference.md:1963]:
```suggestion
| deserializer | Returns a deserealized object from a string.                                   |
```

File: docs/api-reference.md [hunk 4]
```diff
@@ -1730,3 +1730,235 @@ Resolves a relative path to an absolute URL within the application.
...
+```
+
+The `params` object exposes the following API:
+
+| Field Name   | Description                                                                      |
+| ------------ | -------------------------------------------------------------------------------- |
+| fetch        | A JavaScript fetch function that you can use to perform arbitrary HTTP requests. |
+| info         | An object that describes the current request.                                    |
+| init         | An object that is the fetch function's `init` parameter.                         |
+| repeat       | A function that you can call to repeat the current request.                      |
```
Comment (dirk-pieterse) [inline docs/api-reference.md:1960]:
```suggestion
| repeat       | A function you can call to repeat the current request.                      |
```

File: docs/api-reference.md [hunk 5]
```diff
@@ -1730,3 +1730,235 @@ Resolves a relative path to an absolute URL within the application.
...
+
+#### Implement Custom Middleware
+
+You can define custom middleware as follows:
+
+```js
+const myMiddleware = async (
+  options, // Options passed to the factory function.
+  response, // The second argument is either a response or error.
+  params // Contains API that you can use in your middleware implementation. See the API table below.
```
Comment (dirk-pieterse) [inline docs/api-reference.md:1943]:
```suggestion
  params // Contains API you can use in your middleware implementation. See the API table below.
```

File: docs/api-reference.md [hunk 6]
```diff
@@ -1730,3 +1730,235 @@ Resolves a relative path to an absolute URL within the application.
...
+
+Validates the response and retries if the validation fails. This allows you to check whether the response contains the latest data and, if it does not, wait for a Read Model to update.
+
+Initialized by the `createWaitForResponseMiddleware` factory function.
+
+The `waitForResponse` middleware takes the following options:
+
+| Option Name | Description                                                          |
+| ----------- | -------------------------------------------------------------------- |
+| attempts    | The number of retries on validation error.                           |
```
Comment (dirk-pieterse) [inline docs/api-reference.md:1894]: "retries on validation error" What do you mean when you say 'on' validation error?

File: docs/api-reference.md [hunk 7]
```diff
@@ -1730,3 +1730,235 @@ Resolves a relative path to an absolute URL within the application.
...
+| debug       | If set to `true`, the middleware logs errors to the browser console. |
```
Comment (dirk-pieterse) [inline docs/api-reference.md:1895]:
```suggestion
| debug       | If set to `true`, the middleware logs errors in the browser console. |
```

File: docs/api-reference.md [hunk 8]
```diff
@@ -1730,3 +1730,235 @@ Resolves a relative path to an absolute URL within the application.
...
+| period      | A time period to wait between retries specified in milliseconds.     |
```
Comment (dirk-pieterse) [inline docs/api-reference.md:1896]:
```suggestion
| period      | The time between retries specified in milliseconds.     |
```

File: docs/api-reference.md [hunk 9]
```diff
@@ -1730,3 +1730,235 @@ Resolves a relative path to an absolute URL within the application.
...
+    if (err) {
+      console.warn(`Error while sending command: ${err}`)
+    }
+  }
+)
+```
+
+##### waitForResponse
+
+Validates the response and retries if the validation fails. This allows you to check whether the response contains the latest data and, if it does not, wait for a Read Model to update.
```
Comment (dirk-pieterse) [inline docs/api-reference.md:1886]:
```suggestion
Validates the response and retries if validation fails. This allows you to check whether the response contains the latest data or wait for the Read Model to update.
```

File: docs/api-reference.md [hunk 10]
```diff
@@ -1730,3 +1730,235 @@ Resolves a relative path to an absolute URL within the application.
...
+Retries the request if the server responds with an error. Initialized by the `createRetryOnErrorMiddleware` factory function.
+
+The `retryOnError` middleware takes the following options:
+
+| Option Name | Description                                                          |
+| ----------- | -------------------------------------------------------------------- |
+| attempts    | The number of retries on error.                                      |
+| errors      | An array of error codes that are allowed to trigger a retry.         |
+| debug       | If set to `true`, the middleware logs errors to the browser console. |
+| period      | A time period to wait between retries specified in milliseconds.     |
```
Comment (dirk-pieterse) [inline docs/api-reference.md:1849]: See the comment below

File: docs/api-reference.md [hunk 11]
```diff
@@ -1730,3 +1730,235 @@ Resolves a relative path to an absolute URL within the application.
...
+
+##### retryOnError
+
+Retries the request if the server responds with an error. Initialized by the `createRetryOnErrorMiddleware` factory function.
+
+The `retryOnError` middleware takes the following options:
+
+| Option Name | Description                                                          |
+| ----------- | -------------------------------------------------------------------- |
+| attempts    | The number of retries on error.                                      |
```
Comment (dirk-pieterse) [inline docs/api-reference.md:1846]: Why 'on' error?

File: docs/api-reference.md [hunk 12]
```diff
@@ -1730,3 +1730,235 @@ Resolves a relative path to an absolute URL within the application.
...
+    }
+  }
+)
+```
+
+##### retryOnError
+
+Retries the request if the server responds with an error. Initialized by the `createRetryOnErrorMiddleware` factory function.
+
+The `retryOnError` middleware takes the following options:
```
Comment (dirk-pieterse) [inline docs/api-reference.md:1842]: How can something take an option?

File: docs/api-reference.md [hunk 13]
```diff
@@ -1730,3 +1730,235 @@ Resolves a relative path to an absolute URL within the application.
...
+| ------------------- | ----------------------------------------------------------- |
+| [parseResponse]()   | Deserializes the response data if it contains valid JSON.   |
+| [retryOnError]()    | Retries the request if the server responds with an error.   |
+| [waitForResponse]() | Validates the response and retries if the validation fails. |
+
+##### parseResponse
+
+Deserializes the response data if it contains valid JSON. If the data is not JSON, the original string is kept. Initialized by the `createParseResponseMiddleware` factory function.
+
+This middleware takes no options. You can add it to a request as shown below:
```
Comment (dirk-pieterse) [inline docs/api-reference.md:1819]: How can something 'take' options?

File: docs/api-reference.md [hunk 14]
```diff
@@ -1730,3 +1730,235 @@ Resolves a relative path to an absolute URL within the application.
...
+
+#### Available Middlewares
+
+This section lists request middleware included into the resolve-client package. The following middleware is available:
+
+| Name                | Description                                                 |
+| ------------------- | ----------------------------------------------------------- |
+| [parseResponse]()   | Deserializes the response data if it contains valid JSON.   |
+| [retryOnError]()    | Retries the request if the server responds with an error.   |
+| [waitForResponse]() | Validates the response and retries if the validation fails. |
```
Comment (dirk-pieterse) [inline docs/api-reference.md:1813]:
```suggestion
| [waitForResponse]() | Validates the response and retries if validation fails. |
```

Comment (dirk-pieterse) [reply docs/api-reference.md:1813]: Retries what?

File: docs/api-reference.md [hunk 15]
```diff
@@ -1730,3 +1730,235 @@ Resolves a relative path to an absolute URL within the application.
...
+    middleware: {
+      response: [
+        // An array of middleware that runs on server response
+        createMyResponseMiddleware({
+          // Middleware options
+        }),
+        ...
+      ],
+      error: [
+        // An array of middleware that runs on server error
```
Comment (dirk-pieterse) [inline docs/api-reference.md:1758]: Do you mean 
```suggestion
        // An array of middleware that runs when there is a server error
```

**Review by dirk-pieterse — APPROVED** [review]

File: docs/api-reference.md (hunk 12 above)
Comment (EugeniyBurmistrov) [reply docs/api-reference.md:1842]: Changed to 'has the following options'

File: docs/api-reference.md (hunk 13 above)
Comment (EugeniyBurmistrov) [reply docs/api-reference.md:1819]: Changed to "has no options"

File: docs/api-reference.md (hunk 10 above)
Comment (EugeniyBurmistrov) [reply docs/api-reference.md:1849]: Changed according to your edit in other section:
| debug       | If set to `true`, the middleware logs errors in the browser console. |
| period       | The time between retries specified in milliseconds.                         |

File: docs/api-reference.md (hunk 6 above)
Comment (EugeniyBurmistrov) [reply docs/api-reference.md:1894]: Changed to "The number of retries if validation fails.  "

File: docs/api-reference.md (hunk 11 above)
Comment (EugeniyBurmistrov) [reply docs/api-reference.md:1846]: Changed to "The number of retries if the server responds with an error. "

Comment (dirk-pieterse) [reply docs/api-reference.md]: Ok

File: docs/api-reference.md (hunk 12 above)
Comment (dirk-pieterse) [reply docs/api-reference.md]: Ok

File: docs/api-reference.md (hunk 13 above)
Comment (dirk-pieterse) [reply docs/api-reference.md]: Ok

**Pull request merged**